### PR TITLE
Redirect on missing trailing slash, fixes #72

### LINKED ISSUE #72: Redirect with trailing slashes
If the user requests a page `/foo/bar` but sfz (with `-r`) renders `/foo/bar/index.html`, it should first redirect the user permanently to `/foo/bar/`.

This is consistent with most HTTP server implementationns. Lack of this redirection may lead to relative path resolution issues, especially when serving generated docs (e.g. rustdoc).

### PR DESCRIPTION
**Pull request by SOF3**

### CODE REVIEW

**Comment by SOF3** [comment]
How come I forget to pull upstream every time -_-

File: src/server/res.rs [hunk 1]
```diff
@@ -9,11 +9,27 @@
...
 
-use headers::{ContentLength, HeaderMapExt};
+use headers::{ContentLength, Header, HeaderMapExt, HeaderValue, Location};
 use hyper::StatusCode;
 
 use crate::server::Response;
 
+/// Generate 304 NotModified response.
+pub fn moved_permanently(mut res: Response, src: &str) -> Response {
+    *res.status_mut() = StatusCode::MOVED_PERMANENTLY;
```
Comment (weihanglo) [inline src/server/res.rs:19]: I wonder shall we use "302 Found" instead? Since some scenarios of sfz are for temporary use, and we do not want Client to cache the redirection indefinitely.

File: src/server/res.rs [hunk 2]
```diff
@@ -9,11 +9,27 @@
...
+    let location = match HeaderValue::from_str(src) {
+        Ok(location) => location,
+        Err(err) => panic!("{} cannot be formatted into a Location header: {}", src, err),
+    };
+    let location = match Location::decode(&mut std::iter::once(&location)) {
+        Ok(location) => location,
+        Err(err) => panic!("{} cannot be formatted into a Location header: {}", src, err),
+    };
```
Comment (weihanglo) [inline src/server/res.rs:27]: This can be reduced into one match expression.

```suggestion
    let location = match Location::decode(&mut HeaderValue::from_str(src).iter()) {
        Ok(location) => location,
        Err(err) => panic!("{} cannot be formatted into a Location header: {}", src, err),
    };
```

File: src/server/res.rs [hunk 3]
```diff
@@ -9,11 +9,27 @@
...
+/// Generate 304 NotModified response.
+pub fn moved_permanently(mut res: Response, src: &str) -> Response {
+    *res.status_mut() = StatusCode::MOVED_PERMANENTLY;
+    let location = match HeaderValue::from_str(src) {
+        Ok(location) => location,
+        Err(err) => panic!("{} cannot be formatted into a Location header: {}", src, err),
+    };
+    let location = match Location::decode(&mut std::iter::once(&location)) {
+        Ok(location) => location,
+        Err(err) => panic!("{} cannot be formatted into a Location header: {}", src, err),
```
Comment (weihanglo) [inline src/server/res.rs:26]: It's not appropriate to panic directly since it's not an unexpected error. Perhaps we can return an `Result::Err` and let caller deal with it (maybe extend `internal_server_error` to show the info).

**Review by weihanglo — COMMENTED** [review]
Sorry for the late review and thanks for your contribution. Generally it's a great enhancement, just need some changes

**Comment by weihanglo** [comment]
Could you please run `cargo fmt`? Thanks!

File: src/server/res.rs [hunk 4]
```diff
@@ -9,11 +9,27 @@
 //! Response factory functions.
 //!
 
-use headers::{ContentLength, HeaderMapExt};
+use headers::{ContentLength, Header, HeaderMapExt, HeaderValue, Location};
 use hyper::StatusCode;
 
 use crate::server::Response;
 
+/// Generate 304 NotModified response.
```
[the remark this answers is not in the document]
Comment (SOF3) [reply src/server/res.rs:17]: Actually 301. Will update that shortly.

File: src/server/res.rs (hunk 2 above)
Comment (SOF3) [reply src/server/res.rs:27]: This doesn't report errors from the from_str call, does it?

Comment (weihanglo) [reply src/server/res.rs:27]: Ah. Yep. You're right 👍🏾

File: src/server/res.rs (hunk 1 above)
Comment (SOF3) [reply src/server/res.rs:19]: It's probably unlikely that a directory and a file have the same name, right?

File: src/server/res.rs (hunk 3 above)
Comment (SOF3) [reply src/server/res.rs]: The location string, for the use case in this PR, originates from the client request. If the location string is malformed, this probably means the client intentionally sent some characters such as `%00` (is this even possible?) to trigger this case. Perhaps `400 Bad Request` would suffice?

File: src/server/res.rs (hunk 1 above)
Comment (weihanglo) [reply src/server/res.rs:19]: Yes. But the content you serve may change. We do not want client to cache 301 forever.

File: src/server/res.rs (hunk 3 above)
Comment (weihanglo) [reply src/server/res.rs]: If we handle the error in the `moved_permanently` caller side, we can expect that we pass in a `req.uri().path()`, which is a valid percent-encoded str. According to [RFC2616 Section-4.2], header field-content can be arbitrary `TEXT`, so passing str with percent-encoding, which is a subset of US-ASCII, is fine. 

As you mentioned, the **NUL**  is not a reserved nor unreserved character listed in [RFC3986 Section-2]. It will be encoded to `%00`, which is also a valid value that fits header's field-content. 

In conclusion, `req.uri().path()` gives us a valid percent-encoded value to pass into header. If we still get an error, it should be a problem from our side (or from headers crate 😆). So, personally 500 is fine.

[RFC2616 Section-4.2]: https://datatracker.ietf.org/doc/html/rfc2616#section-4.2
[RFC3986 Section-2]: https://datatracker.ietf.org/doc/html/rfc3986#section-2

Comment (weihanglo) [reply src/server/res.rs]: I got what you mean and a request without percent-encoding URI really worth a bad requset. My contention here is I trust hyper does that for us but I am not 100% sure. And I am also ok with 400. 😄

**Comment by weihanglo** [comment]
Thank you for the PR. Unfortunately I have no time working on it. See #108.
